plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,632 @@
+  depGraph :: AM.AdjacencyMap PLC.Unique
+  depGraph = AM.induceJust .
+             AM.gmap (\case Variable u -> Just u;
+                            -- we remove Root because we do not care about it
+                            Root -> Nothing)
+             $ runTermDeps pir
+
+  -- | the dependency graph as before, but with datatype-bind nodes merged/reduced under the "principal" node, See Note [Principal].
+  reducedDepGraph :: AM.AdjacencyMap PLC.Unique
+  reducedDepGraph = M.foldr (\rhs accGraph ->
+                               let ids = rhs^..rhsBinding.bindingIds
+                               in case ids of
+                                 -- A lot of binds are termbinds/typebinds with no vertices to merge.
+                                 -- This optimizes all these cases of termbinds/typebinds to avoid traversing in O(n) the graph
+                                 -- looking for "possible" merges, because there are none to be performed
+                                 [_nonDatatypeBind] -> accGraph
+                                 _   ->  AM.mergeVertices (`S.member` S.fromList ids) (rhs^.rhsBinding.principal) accGraph
+
+                            ) depGraph rhsTable
+
+  -- |take the strongly-connected components of the reduced dep graph, because it may contain loops (introduced by the LetRecs)
+  -- topologically sort these sccs, since we rely on linear (sorted) scoping in our 'genLets' code generation
+  topSortedSccs :: [NS.NESet PLC.Unique]
+  topSortedSccs =
+    let allLets = M.keysSet rhsTable
+    in mapMaybe (\scc ->  NS.nonEmptySet $
+                  -- we are not interested in graph structure anymore
+                  -- make sure that scc contain only to-be floated lets (i.e. no lambdas and no fixed strictNonValue lets)
+                  AMN.vertexSet scc `S.intersection` allLets
+                )
+       . fromRight (error "Cycle detected in the scc-graph. This shouldn't happen in the first place.") . AM.topSort $ AM.scc reducedDepGraph
+
+  -- | Groups a given set of lets into one or more multilets and wraps these multilets around a term.
+  -- The grouping is done through the strongly-connected components
+  -- The input lets are not sorted w.r.t. linear scoping, so this function uses the topological-sort of these SCCs,
+  -- to figure out the correct (dependent/linear) order in which to generate these new multilets.
+  --
+  -- The resulting term is wrapped with linear-scope-sorted LetRecs and LetNonRecs (interspersed between each other because letnonrec and letrec syntax cannot be combined)
+  -- Example: `let {i = e, ...} in let rec {j = e, ...} in let rec {...} in let {...} in ..... in originalTerm`
+  genLets :: Maybe (NS.NESet PLC.Unique) -- ^ all the let identifiers to wrap around this term
+          -> FloatData -- ^ the remaining data to be floated
+          -> State [NS.NESet PLC.Unique] (Term tyname name uni a -> Term tyname name uni a) -- ^ a wrapper function of term
```

Comment:
That's how I had it in a previous commit. I can try to explain why I did this change; it is essentially changing `genLets` from a monad-action to an applicative style. This way I am sure that the `State` cannot be influenced by the input-term. It is just to make more safe the `genLets`. If you don't like it I can revert it, and use `=<<` instead of `<*>` when I apply `genLets`

Reply:
It's not *wrong*, I guess it's just surprising. If you did it for a reason then I don't mind!